File: twosum.py
def two_sum(nums, target):
    """
    Find two numbers in the list that add up to the target.
    
    Args:
        nums: List of integers
        target: Target sum
    
    Returns:
        List of two indices where nums[i] + nums[j] == target
        Returns empty list if no valid pair exists
    """
    seen = {}
    
    for i, num in enumerate(nums):
        complement = target - num
        
        if complement in seen:
            return [seen[complement], i]
        
        seen[num] = i
    
    return []

File: src/twosum.cpp
#include "twosum.h"
#include <unordered_map>

std::vector<int> two_sum(const std::vector<int>& nums, int target) {
    std::unordered_map<int, int> seen;
    seen.reserve(nums.size());

    for (int i = 0; i < static_cast<int>(nums.size()); ++i) {
        int complement = target - nums[i];
        auto it = seen.find(complement);
        if (it != seen.end()) {
            return {it->second, i};
        }
        // map value to index (first occurrence only)
        if (seen.find(nums[i]) == seen.end()) {
            seen[nums[i]] = i;
        }
    }

    return {};
}
